src/gardenlinux/s3/s3_object_index.py

```diff
@@ -0,0 +1,140 @@
+# -*- coding: utf-8 -*-
+
+"""
+S3 object index with flavors filtering
+"""
+
+import base64
+import json
+import logging
+import os
+import subprocess
+import time
+import yaml
+from typing import Any, Optional
+
+from ..flavors.parser import Parser
+from ..logger import LoggerSetup
+from .bucket import Bucket
+
+
+class S3ObjectIndex(object):
+    """
+    S3 object index class with flavors filtering capabilities
+
+    :author:     Garden Linux Maintainers
+    :copyright:  Copyright 2024 SAP SE
+    :package:    gardenlinux
+    :subpackage: s3
+    :since:      0.9.0
+    :license:    https://www.apache.org/licenses/LICENSE-2.0
+                 Apache License, Version 2.0
+    """
+
+    def __init__(
+        self,
+        bucket_name: str,
+        endpoint_url: Optional[str] = None,
+        s3_resource_config: Optional[dict[str, Any]] = None,
+        logger: Optional[logging.Logger] = None,
+    ):
+        """
+        Constructor __init__(S3ObjectIndex)
+
+        :param bucket_name: S3 bucket name
+        :param endpoint_url: S3 endpoint URL
+        :param s3_resource_config: Additional boto3 S3 config values
+        :param logger: Logger instance
+
+        :since: 0.9.0
+        """
+
+        if logger is None or not logger.hasHandlers():
+            logger = LoggerSetup.get_logger("gardenlinux.s3")
+
+        self._bucket = Bucket(bucket_name, endpoint_url, s3_resource_config)
+        self._logger = logger
+
+    def get_index(
+        self,
+        prefix: str,
+        cache_file: Optional[str] = None,
+        cache_ttl: int = 3600,
+    ) -> dict[str, Any]:
+        """
+        Get and cache S3 objects with an indexed list of objects.
+
+        :param prefix:     Prefix for S3 objects
+        :param cache_file: Path to cache file (optional, enables caching when provided)
+        :param cache_ttl:  Cache time-to-live in seconds
+
+        :returns: Dictionary containing 'index' and 'artifacts' keys
+
+        :since: 0.9.0
+        """
+
+        self._logger.debug(f"Getting object index for prefix: {prefix}")
+
+        # Fetch directly if no caching
+        if cache_file is None:
+            artifacts = [
+                s3_object.key
+                for s3_object in self._bucket.objects.filter(Prefix=prefix).all()
+            ]
+            self._logger.debug(f"Fetched {len(artifacts)} artifacts without caching")
+            return {"index": self._build_index(artifacts), "artifacts": artifacts}
+
+        # Check cache
+        index_file = cache_file + ".index.json"
```

Comment:
I do see an issue with `prefix` not being part of the cache file name or some kind of metadata. Whatever you give as `prefix` it would still use `cache_file` no matter if the source file of the cached file equals to the later execution of this method.

Beside that, as the cache file parameter is optional this functionality should be splitted:
- 1. Call underlying `list_uploaded_artifacts()` with a filter parameter containing `Prefix`.
- 2a. Generate cache with help of `cache_list_of_uploaded_artifacts()`
- 2b. Return result if `cache_file` is `None`.

As the index structure is easily rebuild I think there is no performance advantage in caching it.